### PR TITLE
UI: Rekisteröitymisen pohja

### PR DESCRIPTION
**Pull request by jkaariain**
- react-router-dom lisätty reititystä varten
- react-bootstrap lisätty
- react-icons lisätty
- formik lisätty (lomakkeen validointia varten)

Luotu barebones-versio rekisteröintikomponentista. Luo käyttäjä -painike ei tee mitään muuta kuin päivittää sivun. Päänäkymässä on linkki rekisteröintisivulle, josta pääsee takaisin painamalla Takaisin-linkkiä lomakkeen yläpuolella.

### CODE REVIEW

File: src/components/SignUpForm.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
+import React, { useState } from 'react';
+import Form from 'react-bootstrap/Form';
+import Button from 'react-bootstrap/Button';
+
+function SignUpFormHeader() {
+  return <h3>Luo uusi käyttäjä</h3>;
+}
```
Comment (Akzuu) [inline src/components/SignUpForm.tsx]: I do personally prefer arrow functions over this. No need to change though

File: src/components/SignUpForm.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+function SignUpForm() {
+  const [values, setValues] = useState({
+    firstName: '',
+    lastName: '',
+    userId: '',
```
Comment (Akzuu) [inline src/components/SignUpForm.tsx]: `userId` -> `username`.

Id would imo be something that is randomly generated in the database

File: src/components/SignUpForm.tsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    password: '',
+    passwordConfirm: '',
```
Comment (Akzuu) [inline src/components/SignUpForm.tsx]: Wouldn't `undefined` be clearer as the initialization value?

File: src/components/SignUpForm.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  });
+
+  const handleFirstNameInputChange = (
```
Comment (Akzuu) [inline src/components/SignUpForm.tsx]: Add `useCallback` to avoid unnecessary renders https://reactjs.org/docs/hooks-reference.html#usecallback

Comment (Akzuu) [reply src/components/SignUpForm.tsx]: Same for each of the functions below

**Review by Akzuu — CHANGES_REQUESTED** [review]

**Comment by Akzuu** [comment]
There could also be tests to validate that form validation works (gives errors if a field is left empty etc)

File: src/components/SignUpForm.tsx (hunk 3 above)
Comment (jkaariain) [reply src/components/SignUpForm.tsx]: "string is not assignable to a type undefined."
image

**Comment by jkaariain** [comment]
> There could also be tests to validate that form validation works (gives errors if a field is left empty etc)

There is a "please fill out this field" tooltip sent by the browser if the user attempts to send a form with one or more empty fields. Also, the verification is usually done server-side (backend sends error messages, which are displayed by the input fields.)

**Comment by Akzuu** [comment]
> > There could also be tests to validate that form validation works (gives errors if a field is left empty etc)
> 
> There is a "please fill out this field" tooltip sent by the browser if the user attempts to send a form with one or more empty fields. Also, the verification is usually done server-side (backend sends error messages, which are displayed by the input fields.)

Always asking backend for validation seems slow and inefficient, in my opinion both frontend and backend should do basic validation.

Anyways, should we a have test that checks if the popup is shown?

File: src/components/SignUpForm.tsx [hunk 5]
```diff
@@ -1,69 +1,64 @@
...
-}
+};
 
-function SignUpForm() {
+const SignUpForm = () => {
   const [values, setValues] = useState({
     firstName: '',
     lastName: '',
-    userId: '',
+    username: '',
```
Comment (Akzuu) [inline src/components/SignUpForm.tsx:13]: I started to wonder in the shower, if we really even need a `username`. Where would that be used? We probably want to use email as the login identifier, since that is one less thing for the users to remember.

**Review by Akzuu — APPROVED** [review]

File: src/components/SignUpForm.tsx (hunk 5 above)
Comment (Akzuu) [reply src/components/SignUpForm.tsx:13]: So maybe change this again to `email` if this approach seems correct to also you. 

In the future we have to generate some kind of check validate email addresses that they actually exist, but we can think about that later on.

**Comment by ilesoft** [comment]
> > There could also be tests to validate that form validation works (gives errors if a field is left empty etc)
> 
> There is a "please fill out this field" tooltip sent by the browser if the user attempts to send a form with one or more empty fields. Also, the verification is usually done server-side (backend sends error messages, which are displayed by the input fields.)

There should be both client-side validation and server-side validation and verification.

* Client-side to tell user not to submit phone number as a cylinder volume or pet name as an email.
* Server-side to send client _400 You failed_ response or that there is already an account that with that email etc.

Anyway, not in the scope of this PR.

**Review by ilesoft — APPROVED** [review]
